### PR TITLE
Consolidate numpy and dask vindex handling

### PR DESCRIPTION
**Pull request by shoyer**

### CODE REVIEW

File: xarray/core/nputils.py [hunk 1]
```diff
@@ -79,3 +81,55 @@ def array_ne(self, other):
...
+        return (), ()
+
+    non_slices = [k for k in key if not isinstance(k, slice)]
+    ndim = len(np.broadcast(*non_slices).shape)
+    mixed_positions = advanced_index_positions[0] + np.arange(ndim)
+    vindex_positions = np.arange(ndim)
+    return mixed_positions, vindex_positions
+
+
+class VectorizedIndex(object):
```
Comment (fujiisoup) [inline xarray/core/nputils.py]: As it is an array-wrapper, I think the name should be something like `NumpyVectorizedIndexingAdapter`, like `NumpyIndexingAdapter` or `DaskIndexingAdapter`.

File: xarray/core/variable.py [hunk 2]
```diff
@@ -463,77 +471,67 @@ def _nonzero(self):
...
+        for i, (dim, value) in enumerate(zip(self.dims, key)):
+            if isinstance(value, slice):
+                if dim in variable_dims:
+                    # We only convert slice objects to variables if they share
+                    # a dimension with at least one other variable. Otherwise,
+                    # we can equivalently leave them as slices and transpose the
+                    # result. This is significantly faster/more efficient for
+                    # most array backends.
+                    values = np.arange(*value.indices(self.sizes[dim]))
+                    variables.insert(i, Variable((dim,), values))
```
Comment (fujiisoup) [inline xarray/core/variable.py]: I am wondering if this line is correct.
```python
v = Variable(['w', 'x', 'y', 'z'], [[[[1, 2, 3], [4, 5, 6]]]])
ind = Variable(['y'], [0])
v_new = v[ind, :, 1:2, 2]
```
raises
'IndexError: index 2 is out of bounds for axis 2 with size 2'

**Review by fujiisoup — COMMENTED** [review]
Thanks for the PR. I have some comments here

File: xarray/core/nputils.py (hunk 1 above)
Comment (shoyer) [reply xarray/core/nputils.py]: Agreed, changed.

File: xarray/core/variable.py (hunk 2 above)
Comment (shoyer) [reply xarray/core/variable.py]: Good catch! Fixed (I think)

**Pull request merged**